jenkins/printJarVersion.sh

```diff
@@ -24,9 +25,9 @@ function print_ver(){
     
     if [[ "$VERSION" == *"-SNAPSHOT" ]]; then
         PREFIX=${VERSION%-SNAPSHOT}
-        TIMESTAMP=`grep -oP '(?<=timestamp>)[^<]+' < $REPO/maven-metadata-$SERVER_ID.xml`
-        BUILD_NUM=`grep -oP '(?<=buildNumber>)[^<]+' < $REPO/maven-metadata-$SERVER_ID.xml`
-        echo $TAG=$PREFIX-$TIMESTAMP-$BUILD_NUM$SUFFIX
+        # List the latest SNAPSHOT jar file in the maven repo
+        FILE_NAME=`ls -t $REPO/$PREFIX-[0-9]*$SUFFIX | head -1 | xargs basename`
```

Comment:
Have you tested this when there are no timestamped versions of the cudf snapshot jar?  (e.g.: developer always builds cudf from source themselves.)

It's annoying that we're jumping through all these hoops to get what Maven already knows.  It would be much better if we could get Maven itself to emit this information.  Unfortunately it doesn't look like Maven has any easy way to do this.  I did find a Maven extension that can record the dependency info at https://github.com/khmarbaise/deployment-recorder-extension but it looks like it only records events when an artifact is deployed, not resolved, so I think it would only record the information we need if the cudf artifact was downloaded during that particular build.

Reply:
@jlowe Unfortunately it doesn't look like Maven has any easy way to do this.  --- exactly.
I know there is a file `_remote.repositories` which has all the record downloaded from remote as below. Still we need to parse this file to get the latest time stamp:

cudf-0.17-20201110.090227-19-cuda10-2.jar>apache.snapshots=
cudf-0.17-20201110.070212-17-cuda10-2.jar>apache.snapshots=
cudf-0.17-20201103.102820-14.pom>apache.snapshots=
cudf-0.17-20201103.102820-14-cuda10-1.jar>temp=
cudf-0.17-20201110.090227-19-cuda10-1.jar>apache.snapshots=
cudf-0.17-20201110.090005-20-cuda10-1.jar>snapshots-repo=
cudf-0.17-20201103.102820-14-cuda10-1.jar>apache.snapshots=

Reply:
@jlowe Have you tested this when there are no timestamped versions of the cudf snapshot jar? (e.g.: developer always builds cudf from source themselves.) --- Do you mean that we install cudf dependency into the .m2 dir? 
I did not test this case. Maybe there is a problem. In case we installed cudf into .m2, instead of getting from remote, then the latest jars not from remote. 

So the real dependency should be the local install one? Or we still think that the remote one the the correct dependency?

Reply:
Do we really need to gather the dependency version in local building?
How about we do this printJarVersion.sh only when our nightly or release building?
It can avoid the developer building and make the life easy.

Reply:
Just test that when we install cudf locally by mvn install cudf-0.17*.jar, it only overwrite the cudf-0.17-SNAPSHOT-cuda*.jar file, along with updating the `maven-metadata-local.xml` file

Reply:
I'm not sure gathering that info is important, but at a minimum we should test it so we know it does not crash. I did a little bit of that and it looks like it will work properly, but I did not actually install the patch and run a build.  I think that is what @jlowe wants because that is what was missed last time and why we had to revert the patch.

Reply:
So if there is only cudf-0.17-SNAPSHOT-cuda*.jar file, this scripts will fail, too.

Reply:
If that is true then we cannot check this in.  We need to detect that.

Reply:
> Do we really need to gather the dependency version in local building?

I believe @tgravescs would like this to work in the off chance that a developer hands a custom build to a user and that user then files a report.  It would be nice to at least indicate in some way the cudf snapshot jar was not an official snapshot from the snapshots repo (e.g.: artifact is just cudf-0.17-SNAPSHOT-cuda10-1.jar without any timestamp, indicating it was a local build).

> ... at a minimum we should test it so we know it does not crash. ... I think that is what @jlowe wants

Exactly.  Putting something useful in the properties for a local developer build would be great, but at a minimum this should not crash the build, local or otherwise, when everything is working normally.

Reply:
Ok, I'll check to only fail the Jenkins build, and NOT fail the local build.
@tgravescs Need we set `cudfVer=cudf-0.17-SNAPSHOT-cuda10-1.jar` dependency info for the local build, if we install the dev built cudf jars?

Reply:
I think we can know the local building just from the dependency info is missing. And the cudf version should be included in pom file.

Reply:
Note that there can be some snapshot files intermixed with local build files in the maven repository.  For example from my local repo:
```
$ ls -lrt ~/.m2/repository/ai/rapids/cudf/0.17-SNAPSHOT
total 613400
-rw-rw-r-- 1 jlowe jlowe     23109 Oct 28 11:53 cudf-0.17-20201028.162547-8.pom
-rw-rw-r-- 1 jlowe jlowe        40 Oct 28 11:53 cudf-0.17-20201028.162547-8.pom.sha1
-rw-rw-r-- 1 jlowe jlowe         0 Oct 28 11:53 cudf-0.17-20201028.162547-8-cuda10-1.jar.part.lock
-rw-rw-r-- 1 jlowe jlowe  39518965 Oct 28 11:53 cudf-0.17-20201028.162547-8-cuda10-1.jar.part
-rw-rw-r-- 1 jlowe jlowe     23109 Nov  5 08:23 cudf-0.17-20201105.090410-17.pom
-rw-rw-r-- 1 jlowe jlowe        40 Nov  5 08:23 cudf-0.17-20201105.090410-17.pom.sha1
-rw-rw-r-- 1 jlowe jlowe 214535283 Nov  5 08:24 cudf-0.17-20201105.090410-17-cuda10-1.jar
-rw-rw-r-- 1 jlowe jlowe        40 Nov  5 08:24 cudf-0.17-20201105.090410-17-cuda10-1.jar.sha1
-rw-rw-r-- 1 jlowe jlowe      1406 Nov 11 08:32 maven-metadata-snapshots-repo.xml
-rw-rw-r-- 1 jlowe jlowe        40 Nov 11 08:32 maven-metadata-snapshots-repo.xml.sha1
-rw-rw-r-- 1 jlowe jlowe       315 Nov 11 08:32 resolver-status.properties
-rw-rw-r-- 1 jlowe jlowe     23203 Nov 11 08:32 cudf-0.17-20201111.093609-21.pom
-rw-rw-r-- 1 jlowe jlowe        40 Nov 11 08:32 cudf-0.17-20201111.093609-21.pom.sha1
-rw-rw-r-- 1 jlowe jlowe 242708552 Nov 11 08:35 cudf-0.17-20201111.093609-21-cuda10-1.jar
-rw-rw-r-- 1 jlowe jlowe        40 Nov 11 08:35 cudf-0.17-20201111.093609-21-cuda10-1.jar.sha1
-rw-rw-r-- 1 jlowe jlowe     23203 Nov 11 10:30 cudf-0.17-SNAPSHOT.pom
-rw-rw-r-- 1 jlowe jlowe    356008 Nov 11 15:33 cudf-0.17-SNAPSHOT-tests.jar
-rw-rw-r-- 1 jlowe jlowe 130840131 Nov 11 15:33 cudf-0.17-SNAPSHOT-cuda10-1.jar
-rw-rw-r-- 1 jlowe jlowe       492 Nov 11 15:33 _remote.repositories
-rw-rw-r-- 1 jlowe jlowe       940 Nov 11 15:33 maven-metadata-local.xml
```

Note that there are timestamped snapshots here, but the jar I'm using is not one of them.  If a timestamped snapshot jar is found it should be compared against the "main" jar path (cudf-0.17-SNAPSHOT-cuad10-1.jar in the above example) to know that it indeed is the one being used.

Reply:
@jlowe The "main" jar timestamp is stored in the `maven-metadata-local.xml` file, if we lacally run `maven install`. So when running local build, we only need check timestamps  between in  `maven-metadata-local.xml`(if exists) and in timestamped jars(cudf-0.17-20201111.093609-21-cuda10-1.jar)
-rw-rw-r-- 1 jlowe jlowe       940 Nov 11 15:33 maven-metadata-local.xml

Reply:
> > Do we really need to gather the dependency version in local building?
> 
> I believe @tgravescs would like this to work in the off chance that a developer hands a custom build to a user and that user then files a report. It would be nice to at least indicate in some way the cudf snapshot jar was not an official snapshot from the snapshots repo (e.g.: artifact is just cudf-0.17-SNAPSHOT-cuda10-1.jar without any timestamp, indicating it was a local build).
> 
> > ... at a minimum we should test it so we know it does not crash. ... I think that is what @jlowe wants
> 
> Exactly. Putting something useful in the properties for a local developer build would be great, but at a minimum this should not crash the build, local or otherwise, when everything is working normally.

@jlowe @revans2 @tgravescs @GaryShen2008 
Let's divide it into 2 parts
    We first make sure that: in case of dependency info fetch error, only fails the Jenkins build, and NOT fail the local build. --- this is more import, I think.
    Then let's check if we need to parse the exact local build dependency info, and make another PR to do it. --- This is of low priority, we can optimize it later.

What's your thought?

Reply:
Agree.  We collect dependency-info for nightly and release build, don't collect in local build first, so that local build can pass whatever the environment is.
To collect dependency-info for the local build seems more complicated, and it's low priority.

Reply:
I'm fine with punting on the tricky SNAPSHOT timestamp handling for dev builds for now, but I think it's trivial for us to provide _some_ kind of info for dev builds.  We just need to move the `JENKINS_URL` check to the `SNAPSHOT` check in the script, and then we'll only try to do snapshot timestamp analysis on jenkins builds.  Dev builds will print whatever cudf version was used without timestamp info for snapshots (i.e.: it would just be something like cudf-0.17-SNAPSHOT-cuda10-1 in a dev build).